src/java.base/share/classes/java/lang/reflect/code/bytecode/BytecodeGenerator.java

```diff
@@ -108,7 +104,9 @@ public static <O extends Op & Op.Invokable> MethodHandle generate(MethodHandles.
 
         MethodHandles.Lookup hcl;
         try {
-            hcl = l.defineHiddenClass(classBytes, true);
+            hcl = l.in(l.defineClass(classBytes));
```

Comment:
That's frustrating! but thankfully its easy to work around it

Reply:
Yes, unfortunately hidden classes have some frustrating restrictions.

Reply:
What's the problem here? I recall LambdaMetafactory is designed with hidden class support in mind, that it generates compatible class names if possible.

Since hidden classes allow duplicate names (as it is defined with random suffix), does this break existing usages?

Reply:
Unfortunately this is a different case. We generate a hidden class with lambda implementation method inside and `LambdaMetafactory` could not (and according to the spec should not) find the class.

Reply:
An unfortunate consequence is it is now possible to get linkage errors when generating bytecode for two code model functions with the same name. This causes the `CoreBinaryOpsTest` to fail.

One solution is to append a randomly generated UUID to the class name. Another is to checking if the existing name already resolves to a class, and if so append a count, and repeat until resolution fails. Although that would require some synchronation to protect against races.

Reply:
> Unfortunately this is a different case. We generate a hidden class with lambda implementation method inside and `LambdaMetafactory` could not (and according to the spec should not) find the class.

I believe they can all be of the same nest, but I suppose`InnerClassLambdaMetafactory` would need to patch the hidden target class instance into the constant pool (in theory it could).

It is aware about the target class being hidden when generating the lambda class name:

```
    private static String lambdaClassName(Class<?> targetClass) {
        String name = targetClass.getName();
        if (targetClass.isHidden()) {
            // use the original class name
            name = name.replace('/', '_');
        }
        return name.replace('.', '/') + "$$Lambda";
    }
```

Reply:
`InnerClassLambdaMetafactory` generates the hidden inner class (for example: `lambda_0x00007ff000189400$$Lambda/0x00007ff000117a08`) and then it all fails with:
```
java.lang.ClassNotFoundException: lambda.0x00007ff000189400
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at TestBytecode.consume(TestBytecode.java:291)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at TestBytecode.lambda$testGenerate$14(TestBytecode.java:476)
	at TestBytecode.permutateAllArgs(TestBytecode.java:412)
	at TestBytecode.testGenerate(TestBytecode.java:475)

```

Reply:
I've identified `j/l/i/defineHiddenClass/BasicTest::testLambda` is actually not testing a lambda in a hidden class, but rather a MH in a hidden class.
If I modify `j/l/i/defineHiddenClass/src/Lambda` from `Function<Object, String> f = Object::toString;` to `Function<Object, String> f = o -> o.toString();` the test fails with:
```
java.lang.NoClassDefFoundError: Lambda/0x00007ff00029e800
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:641)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:641)
	at java.base/java.lang.Thread.run(Thread.java:1575)
Caused by: java.lang.ClassNotFoundException: Lambda.0x00007ff00029e800
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
```

Reply:
That seems like a bug to me.

Reply:
`InnerClassLambdaMetafactory` doesn't seem to support hidden classes for `implClass`, but it can easily support them if we allow for hidden `implClass` to only use invokestatic (maybe getstatic and putstatic too) and route such `implClass` to always use condy.